src/peft/utils/safelora.py

```diff
@@ -0,0 +1,229 @@
+
+    """
+
+    base_model_path: str = field(
+        default="meta-llama/Llama-2-7b-hf",
+        metadata={"help": "The path of the base model for obtaining the aligned matrix."},
+    )
+
+    aligned_model_path: str = field(
+        default="TheBloke/Llama-2-7B-Chat-fp16",
+        metadata={"help": "The path of the aligned model for obtaining the aligned matrix."},
+    )
+
+    peft_model_path: str = field(
+        default="LisaSchunke/llama-2-7b-peft-finetuned-20000-dataset",
+        metadata={"help": "The path of the LoRA wieghts and configs."},
+    )
+
+    select_layers_type: str = field(
+        default="number",
+        metadata={"help": "How to select projection layers? options: [threshold, number]."},
+    )
+
+    threshold: float = field(
+        default=0.5,
+        metadata={"help": "The threshold of cosine similarity for selecting projected layers."},
+    )
+
+    num_proj_layers: int = field(
+        default=10,
+        metadata={"help": "The number of projected layers."},
+    )
+
+    devices: str = field(
+        default="cuda",
+        metadata={"help": "Devices are used in SafeLoRA. (cuda or cpu)"},
+    )
+
+    save_weights: bool = field(
+        default=True,
+        metadata={"help": "Replacing and saving SafeLoRA weights to the original LoRA file."},
+    )
+    local_files_only: bool = field(
+        default=False,
+        metadata={"help": "Using for snapshot_download."},
+    )
+
+    dtype: torch.dtype = field(
+        default=torch.bfloat16,
+        metadata={"help": "Data type for model weights, e.g., torch.float32 or torch.bfloat16. If your device is CPU, you should use torch.float32."},
+    )
+
+    def __post_init__(self):
+        if self.base_model_path is None:
+            raise ValueError("base_model_path cannot be None.")
+        if self.aligned_model_path is None:
+            raise ValueError("aligned_model_path cannot be None.")
+        if self.peft_model_path is None:
+            raise ValueError("peft_model_path cannot be None.")
+        if self.devices != 'cuda' and self.dtype != torch.float32:
+            raise ValueError("When using a CPU, please set dtype to torch.float32, as CPUs do not support torch.float16.")
+
+
+def get_aligned_matrix(base_model_path, aligned_model_path, devices, peft_config, configs):
+    """
+    Get projected matrix by following the config (target_modules) from the peft model.
+    The dimensions between the base model's weights and the aligned model's weights should be the same.
+    """
+    sl_align = SafetensorLoader(aligned_model_path, local_files_only=configs.local_files_only)
+    sl_base = SafetensorLoader(base_model_path, local_files_only=configs.local_files_only)
+
+    base_model_parameters = [
+        name for name in sl_base.weight_map.keys() if any(v in name for v in list(peft_config.target_modules))
```

Comment:
This check for `v in list(peft_config.target_modules)` is not always valid. E.g., `target_modules` can be a str in which case a regex match is executed. There can be other edge cases too. I think what should work is to use `LoraModel._check_target_module_exists(peft_config, name)` ([here](https://github.com/huggingface/peft/blob/b1fd97dc3e71d91abf9a8605e8be74ea6bb751c6/src/peft/tuners/lora/model.py#L159)).

Reply:
I have tried to use ```LoraModel._check_target_module_exists(peft_config, name)```.
However, since the model here is not a PEFT model but a complete model, using its name to find the target module will result in not finding it